aepsych/server/sockets.py

```diff
@@ -92,39 +95,90 @@ def __init__(self, port, ip=""):
             )
         else:
             self.socket = socket.create_server(addr)
-        self.conn = None
+
+        self.conn, self.addr = None, None
 
     def close(self):
         self.socket.close()
 
-    def receive(self):
-        # catch the Error and reset the connection
-        while True:
-            try:
-                if self.conn is None:
-                    logger.info("Waiting for connection...")
-                    self.conn, self.addr = self.socket.accept()
-                recv_result = b""
-                while recv_result == b"":
-                    logger.info(f"Connected by {self.addr}, waiting for messages...")
-                    recv_result = self.conn.recv(1024 * 512)  # 512KiB
-                    logger.debug(f"receive : result = {recv_result}")
-                    msg = json.loads(recv_result)
-
-                logger.info(f"Got: {msg}")
-                break
-            except Exception as e:
-                self.conn.close()
-                self.conn, self.addr = None, None
-                logger.info(
-                    "Exception caught while trying to receive a message from the client. "
-                    f"Ignoring message and trying again. The caught exception was: {e}."
-                )
-        return msg
+    def return_socket(self):
+        return self.socket
+
+    def accept_client(self):
+        client_not_connected = True
+        logger.info("Waiting for connection...")
+
+        while client_not_connected:
+            rlist, wlist, xlist = select.select([self.socket], [], [], 0)
+            if rlist:
+                for sock in rlist:
+                    try:
+                        self.conn, self.addr = sock.accept()
+                        logger.info(
+                            f"Connected by {self.addr}, waiting for messages...")
+                        client_not_connected = False
+                    except Exception as e:
+                        logger.info(
+                            f'Connection to client failed with error {e}')
+
+    def receive(self, server_exiting):
+
+        # Set up the logic for reading using select.
+        # Setting the timeout to 0 makes it nonblocking.
+        # Magic Number tweaking can be made with the timeout to cut down CPU Usages vs User Experience
+
+        while not server_exiting:
+            rlist, wlist, xlist = select.select([self.conn], [], [], 0)
+            for sock in rlist:
+                try:
+                    if rlist:
+                        recv_result = sock.recv(1024*512)
```

Comment:
There needs to be some sort of explanation for this magic number. Are you sure there isn't already a built-in function or constant or something that does whatever is being done here? At a minimum, you should put a comment here explaining what is happening.

Reply:
Added more context in function.